Update paper to P3149R10 with small-group wording feedback

ispeters opened the pull request.

ispeters committed:
Update paper number to D3149R10

jesswong committed:
Address Jen's comments

jesswong committed:
Replace nest-data{} initialization with parenthesis
Replace nest-data{} with parentheses

jesswong commented:
Most of Jen's comments were regarding `nest` so I didn't update `nest` since we plan on turning it into prose wording.

Below are Jen's comments and my thoughts:

> C++ does not have "method"s.  You might be talking about
> (non-static) member functions.  Please cleanse the wording accordingly.
> [exec.scope.concepts] p3 (at least) is offending.

Addressed


> [exec.scope.concepts]
> 
> "The async_scope_token<Token> concept..."
> 
> The concept is called "async_scope_token".

Addressed

> Please make a pass through the wording and double-check
> you want various data members to have real names, not
> exposition-only names. I doubt it.
> Example: [exec.nest]  nest-data has various data members
> not marked as exposition-only (and italicized).

I don't quite understand this. A lot of the code examples we have are followed by some prose that includes a variation of "exposition only template class", "is equivalent to the following", or the member function name is italicized. Do we need to also italicize the code examples when it was mentioned earlier that the code is exposition only or that the behavior of the class/member function is "equivalent" to?

> template <async_scope_token Token, sender Sender>
> nest-data(Token, Sender&&) -> nest-data<Token, Sender>;
> 
> Why is that explicit deduction guide needed?  It seems
> to be exactly the same you'd get implicitly from the
> constructor.

Did not address because it will be replaced with prose.

> "nest-data{token, sndr}"
> avoid brace-initialization; use parentheses.

Addressed

> The member impls-for<nest_t>::get-state is initialized with a callable object equivalent to the following lambda:
> 
> []<class Sndr, class Rcvr>(Sndr&& sndr, Rcvr& rcvr) {
> 
> There is no exception-specification; is that intentional?
> Other (existing) get-state lambdas have one.

Maybe I'm missing something, but I don't see other get-state lambdas include explicit exception-specifications?

ispeters commented:
> > The member impls-for<nest_t>::get-state is initialized with a callable object equivalent to the following lambda:
> > []<class Sndr, class Rcvr>(Sndr&& sndr, Rcvr& rcvr) {
> > There is no exception-specification; is that intentional?
> > Other (existing) get-state lambdas have one.
> 
> Maybe I'm missing something, but I don't see other get-state lambdas include explicit exception-specifications?

If you search for "get-state" at https://eel.is/c++draft/exec, there are a number of hits where the provided lambda's `noexcept` clause is explicit. Many are just `noexcept`; some are `noexcept(see below)`. There are also examples of _`get-state`_ having unmentioned `noexcept` clauses. I'm not sure what the implications are, but we should probably have a reason for whatever we choose.

ispeters committed:
Reflow some long lines to be <=120 chars

This diff just reformats some long lines to fit in 120 columns or less.

ispeters committed:
Mostly replace nest-data spec with prose

ispeters committed:
Incorporate feedback from 3/28 session

This change incorporates changes to the specification of _`nest-data`_
suggested during a small group session with @dietmarkuehl, @jesswong,
and @mhoemmen.

jesswong committed:
Address the rest of Jen's comments

jesswong commented on asyncscope.md:
```diff
@@ -3006,13 +3024,13 @@ struct @_impls-for_@<@_join-t_@>: @_default-impls_@ {
 
     static constexpr auto @_get-state_@ =
-        []<class Receiver>(auto&& sender, Receiver& receiver) {
+        []<class Receiver>(auto&& sender, Receiver& receiver) noexcept(/* std::is_nothrow_constructible_v<schedule, decltype(get_scheduler(get_env(receiver))) */) {
```
state construction calls `connect(schedule(get_scheduler(get_env(receiver))), receiver)`. since schedule is a sender itself that has the ability to throw, i thought that checking if `schedule(get_scheduler(get_env(receiver)))` is no_throw constructible might be sufficient. what do you think?

jesswong replied:
Will make this noexcept(false) for now but will ask email chain if wording has a preference

jesswong committed:
Update get-state to be noexcept(false)

ispeters committed:
Partial application of small-group feedback

I'm pushing this midway through applying feedback from the last
small-group session to share the TODOs captured in the section of the
changelog relating to R10.

ispeters committed:
Flesh out nest's get-state

This diff is the result of paring with @jesswong on implementing
`impls-for<nest_t>::get-state` in terms of the newly-defined `release()`
member function on _`nest-data`_. We *think* `get-state` is now correct
and exception-safe; we're still missing an articulation of when
`get-state` is `noexcept`, which is marked in the doc as a TODO.

jesswong committed:
Add noexcept clause for nest_t::get_state

ispeters committed:
Address most feedback from 5/9

This change addresses most of the feedback from the small-group session
on May 9, 2025 and leaves TODOs for what has yet to be addressed.

ispeters committed:
Last-minute changes before mailing deadline

* Incorporate some of Eric Niebler's feedback from the Discord
* Update the revision log
* Attempt to reword [exec.spawn.future] paragraph 4 to better describe
  _`variant_t`_
* Other minor editorial changes

ispeters committed:
Update paper number to P3149R10

R10 has been published so the in-doc number should start with P.
